sui-execution/latest/sui-move-natives/src/crypto/sui_state_proof.rs

```diff
@@ -308,17 +339,34 @@ pub fn sui_state_proof_verify_transaction(
         return Ok(NativeResult::err(cost, INVALID_TX));
     }
 
-    // Ensure the tx is part of the checkpoint
-    let is_valid_checkpoint_tx = checkpoint_contents
-        .iter()
-        .any(|&digest| digest == transaction.effects.execution_digests());
-    if !is_valid_checkpoint_tx {
+    let digests = transaction.effects.execution_digests();
+
+    // Check that transaction digest matches the execution digest
+    if transaction.transaction.digest() != &digests.transaction {
+        return Ok(NativeResult::err(cost, INVALID_TX));
+    }
+
+    // Ensure the digests are in the checkpoint contents
+    if !checkpoint_contents
+        .enumerate_transactions(&summary)
+        .any(|x| x.1 == &digests) {
+            return Ok(NativeResult::err(cost, INVALID_TX));
+    }   
+
+    let events_digest = transaction.events.as_ref().map(|events| events.digest());
+    // Ensure that the execution digest matches the events digest of the passed transaction
+    if !(events_digest.as_ref() == transaction.effects.events_digest()) {
         return Ok(NativeResult::err(cost, INVALID_TX));
     };
 
-    let tx_events = &transaction.events.as_ref().unwrap().data;
+    let tx_events = match transaction.events.as_ref() {
+        Some(events) => &events.data,
+        None => {
+            return Ok(NativeResult::err(cost, INVALID_TX));
+        }
+    };
 
-    let results: Vec<(SuiAddress, Vec<u8>)> = tx_events
+    let results: Vec<(SuiAddress, Vec<Vec<u8>>)> = tx_events
```

Comment:
This will yield a vector in which a single cap_id could appear multiple times with different values. I believe that the way it should be is a `HashMap` (you can always turn it back to a `Vec` before outputting, but this will guarantee uniquness). It doesn't look buggy to me, but it probably is the better approach anyways. If you choose to leave it as-is, just convince yourself there's no bug here.

Reply:
As a dWallet user on Sui, I would expect that if I use the same dWallet cap on Sui to sign several messages within the same PTB, I should be able to submit a state proof for all of them. Using a HashMap would only allow one of these transactions to go through, which is less predictable and creates a worse developer experience IMO.